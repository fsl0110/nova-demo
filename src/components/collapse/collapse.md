### default

```jsx
<Collapse text="Collapse Label">
  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
  voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita
  kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem
  ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod
  tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At
  vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd
  gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
</Collapse>
```

### default expanded

```jsx
<Collapse text="Collapse Label" isOpen={true}>
  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
  voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita
  kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem
  ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod
  tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At
  vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd
  gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
</Collapse>
```

### swiching label text

```jsx
<Collapse text="Collapse Label closed" textOpen="Collapse Label open">
  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
  voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita
  kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem
  ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod
  tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At
  vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd
  gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
</Collapse>
```

### textInline

```jsx
<Collapse text="Collapse Label" textInline={true}>
  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
  voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita
  kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem
  ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod
  tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At
  vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd
  gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
</Collapse>
```

### without icon animation

```jsx
<Collapse text="Collapse Label" iconRotateAnimated={false}>
  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
  voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita
  kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem
  ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod
  tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At
  vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd
  gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
</Collapse>
```

### bigger icon size and other icon

```jsx
<Collapse
  text="Collapse Label"
  iconName="close"
  iconSize="xl"
  iconRotateClosed="45"
  iconRotateOpen="180"
>
  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
  voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita
  kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem
  ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod
  tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At
  vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd
  gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
</Collapse>
```

### width bottom-center positioned close icon button within collapsed content

```jsx
import { Button } from "../button/button";

<Collapse
  text="Collapse Label"
  closeItem={<Button iconName="close" iconSize="l" />}
  closePosition="bottom-center"
>
  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
  voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita
  kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem
  ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod
  tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At
  vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd
  gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
</Collapse>;
```

### width bottom-center positioned close text with icon button within collapsed content

```jsx
import { Button } from "../button/button";

<Collapse
  text="Collapse Label"
  closeItem={
    <Button iconName="close" iconSize="m" iconAlign="after">
      close
    </Button>
  }
  closePosition="bottom-center"
>
  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
  voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita
  kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem
  ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod
  tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At
  vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd
  gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
</Collapse>;
```

### textInline inside text

```jsx
<div>
  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
  voluptua. At vero eos et accusam et justo duo dolores et ea rebum.
  <Collapse text="Collapse Label" textInline={true}>
    Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
    eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
    voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet
    clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit
    amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam
    nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed
    diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet
    clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit
    amet.
  </Collapse>
  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
  voluptua. At vero eos et accusam et justo duo dolores et ea rebum.
</div>
```
