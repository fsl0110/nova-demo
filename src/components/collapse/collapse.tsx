import React, { PureComponent, HTMLAttributes, ReactNode } from "react";
import classNames from "classnames";
import produce from "immer";
import { Icon } from "..";
import { IconRotateProps, IconNameProps, IconSizeProps } from "../icon/icon";

export interface Props extends HTMLAttributes<HTMLDivElement> {
  // Wether the component should be expanded as default. */
  isOpen?: boolean;
  /** Labeltext of the closed collapse. */
  text: string;
  /** Labeltext of the opened collapse. */
  textOpen?: string;
  /** LabelText and LabelIcon left aling e.g as show more text button. */
  textInline: boolean;
  /** Define an icon as hide/show indicator. */
  iconName: IconNameProps;
  /** Define the size of hide/show indicator icon. */
  iconSize: IconSizeProps;
  /** Rotation of the hide/show icon. Default is no rotation. */
  iconRotateClosed?: IconRotateProps;
  /** Rotation of the hide/show icon. Default is 90 degrees. */
  iconRotateOpen: IconRotateProps;
  /** Define wether hide/shown icon toggle should be animated or not. */
  iconRotateAnimated: boolean;
  /** Pass an component as close button. If no one is passed no close button will be shown. */
  closeItem?: ReactNode;
  /** Postion of the passed close button.  */
  closePosition: "bottom-center" | "bottom-right";
  /** Define a label for screenreader */
  ariaLabel?: string;
}

export interface State {
  isOpen: boolean;
}

export class Collapse extends PureComponent<Props, State> {
  state: State = {
    isOpen: this.props.isOpen || false
  };

  static defaultProps: Partial<Props> = {
    iconName: "arrow",
    iconSize: "m",
    iconRotateOpen: "90",
    iconRotateAnimated: true,
    closePosition: "bottom-center"
  };

  toggleIsOpen = () =>
    this.setState(
      produce((draft: State) => {
        draft.isOpen = !draft.isOpen;
      })
    );

  disableIsOpen = () => {
    this.setState(
      produce((draft: State) => {
        draft.isOpen = false;
      })
    );
  };

  render() {
    const {
      text,
      textOpen,
      textInline,
      iconName,
      iconSize,
      iconRotateClosed,
      iconRotateOpen,
      iconRotateAnimated,
      closeItem,
      closePosition,
      className,
      ariaLabel,
      children,
      ...rest
    } = this.props;
    const { isOpen } = this.state;
    const rotate = isOpen ? iconRotateOpen : iconRotateClosed;
    const rootClasses = classNames("collapse", className);
    const labelClasses = classNames("collapse__label", {
      [`collapse__label--text`]: textInline
    });
    const iconClasses = classNames("collapse__label__icon", {
      [`collapse__label__icon--animated`]: iconRotateAnimated
    });
    const contentClasses = classNames("collapse__content", {
      [`collapse__content--closed`]: !isOpen,
      [`collapse__content--open`]: isOpen
    });
    const closeClasses = classNames({
      [`collapse__close`]: !!closeItem,
      [`collapse__close--${closePosition}`]: closeItem && closePosition
    });
    console.log("###", isOpen);
    return (
      <div className={rootClasses} {...rest}>
        <button
          onClick={this.toggleIsOpen}
          aria-label={ariaLabel}
          aria-expanded="false"
          className={labelClasses}
        >
          <div className="label__text">
            {isOpen && textOpen ? textOpen : text}
          </div>
          <Icon
            className={iconClasses}
            name={iconName}
            size={iconSize}
            rotate={rotate}
          />
        </button>
        <div className={contentClasses}>
          {children}
          {closeItem && (
            <span onClick={this.disableIsOpen} className={closeClasses}>
              {closeItem}
            </span>
          )}
        </div>
      </div>
    );
  }
}
