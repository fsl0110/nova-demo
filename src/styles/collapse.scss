.collapse {
}

.collapse__label {
  height: 50px;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.collpase {
  padding-top: 20px;
}

.collapse__content--closed {
  overflow: hidden;
  max-height: 0;
  transition: max-height 0.4s ease-in;
}

.collapse__content--open {
  max-height: 2000px; // bigger than the biggest possible content
  transition: max-height 0.25s ease-out;
}
