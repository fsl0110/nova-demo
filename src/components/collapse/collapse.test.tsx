import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { Collapse, Props } from "./collapse";
import { Button } from "..";

const testCases: Array<[string, Props]> = [
  [
    "Default",
    {
      text: "Collapse Label",
      iconName: "arrow",
      iconSize: "m",
      textInline: false,
      iconRotateOpen: "90",
      iconRotateAnimated: true,
      closePosition: "bottom-center",
      children: "Collapse Content"
    }
  ],
  [
    "textInline",
    {
      text: "Collapse Label",
      iconName: "arrow",
      iconSize: "m",
      textInline: true,
      iconRotateOpen: "90",
      iconRotateAnimated: true,
      closePosition: "bottom-center",
      children: "Collapse Content"
    }
  ],
  [
    "without icon animation",
    {
      text: "Collapse Label",
      iconName: "arrow",
      iconSize: "m",
      textInline: true,
      iconRotateOpen: "90",
      iconRotateAnimated: false,
      closePosition: "bottom-center",
      children: "Collapse Content"
    }
  ],
  [
    "bigger icon size and other icon",
    {
      text: "Collapse Label",
      iconName: "close",
      iconSize: "xl",
      textInline: false,
      iconRotateOpen: "90",
      iconRotateAnimated: true,
      closePosition: "bottom-center",
      children: "Collapse Content"
    }
  ],
  [
    "width bottom-center positioned close icon button within collapsed content",
    {
      text: "Collapse Label",
      iconName: "close",
      iconSize: "xl",
      textInline: false,
      iconRotateClosed: "45",
      iconRotateOpen: "180",
      iconRotateAnimated: false,
      closeItem: <Button iconName="close" iconSize="l" />,
      closePosition: "bottom-center",
      children: "Collapse Content"
    }
  ]
];

describe("<Collapse>", () => {
  testCases.forEach(([item, options]) => {
    it(item, () => {
      expect(toJson(shallow(<Collapse {...options} />))).toMatchSnapshot();
    });
  });
});
