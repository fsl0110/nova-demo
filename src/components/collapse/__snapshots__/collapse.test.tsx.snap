// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapse> Default 1`] = `
<div
  className="collapse"
>
  <button
    aria-expanded="false"
    className="collapse__label"
    onClick={[Function]}
  >
    <div
      className="label__text"
    >
      Collapse Label
    </div>
    <Icon
      className="collapse__label__icon collapse__label__icon--animated"
      name="arrow"
      size="m"
    />
  </button>
  <div
    className="collapse__content collapse__content--closed"
  >
    Collapse Content
  </div>
</div>
`;

exports[`<Collapse> bigger icon size and other icon 1`] = `
<div
  className="collapse"
>
  <button
    aria-expanded="false"
    className="collapse__label"
    onClick={[Function]}
  >
    <div
      className="label__text"
    >
      Collapse Label
    </div>
    <Icon
      className="collapse__label__icon collapse__label__icon--animated"
      name="close"
      size="xl"
    />
  </button>
  <div
    className="collapse__content collapse__content--closed"
  >
    Collapse Content
  </div>
</div>
`;

exports[`<Collapse> textInline 1`] = `
<div
  className="collapse"
>
  <button
    aria-expanded="false"
    className="collapse__label collapse__label--text"
    onClick={[Function]}
  >
    <div
      className="label__text"
    >
      Collapse Label
    </div>
    <Icon
      className="collapse__label__icon collapse__label__icon--animated"
      name="arrow"
      size="m"
    />
  </button>
  <div
    className="collapse__content collapse__content--closed"
  >
    Collapse Content
  </div>
</div>
`;

exports[`<Collapse> width bottom-center positioned close icon button within collapsed content 1`] = `
<div
  className="collapse"
>
  <button
    aria-expanded="false"
    className="collapse__label"
    onClick={[Function]}
  >
    <div
      className="label__text"
    >
      Collapse Label
    </div>
    <Icon
      className="collapse__label__icon"
      name="close"
      rotate="45"
      size="xl"
    />
  </button>
  <div
    className="collapse__content collapse__content--closed"
  >
    Collapse Content
    <span
      className="collapse__close collapse__close--bottom-center"
      onClick={[Function]}
    >
      <Button
        iconName="close"
        iconSize="l"
      />
    </span>
  </div>
</div>
`;

exports[`<Collapse> without icon animation 1`] = `
<div
  className="collapse"
>
  <button
    aria-expanded="false"
    className="collapse__label collapse__label--text"
    onClick={[Function]}
  >
    <div
      className="label__text"
    >
      Collapse Label
    </div>
    <Icon
      className="collapse__label__icon"
      name="arrow"
      size="m"
    />
  </button>
  <div
    className="collapse__content collapse__content--closed"
  >
    Collapse Content
  </div>
</div>
`;
